File: packages/kits/AriaMcpKit/Tests/AriaMCPTests/FactProvenanceTests.swift
// FactProvenanceTests.swift
//
// Tests for two fact-surface fixes:
//
//   Bug C — Host identity injected into ToolDispatcher so rows filed via
//            `moot_file_fact` record which binary hosted the dispatcher
//            (aria-mcp-server, mootx01, etc.). The identity lands in the
//            fact's `addedBy` field and renders as `addedBy=`; `sourceDrawerID`
//            holds a local drawer id or nothing and never a host name.
//
//   Bug D — `moot_fact_search` runs a dark-lane probe when a query is supplied.
//            The probe result is log-side only (recall_provenance removed from
//            payload per COMPOSER-02B). Fact search is purely lexical and
//            returns results regardless of dense-lane availability.

import Testing
import Foundation
import GeniusLocusKit
import LocusKit
import PersistenceKit
import PersistenceKitInMemory
@testable import AriaMCP

// ---------------------------------------------------------------------------
// MARK: - Shared helpers
// ---------------------------------------------------------------------------

/// Extract the text payload from a `textResult` JSONValue.
private func factText(of result: JSONValue) -> String {
    guard case let .object(obj) = result,
          case let .array(content)? = obj["content"],
          case let .object(first)? = content.first,
          case let .string(s)? = first["text"]
    else { return "" }
    return s
}

/// Exercise the public selected-v2 door, not an internal helper.
private func fileFact(
    _ dispatcher: ToolDispatcher, _ arguments: [String: JSONValue]
) async throws -> JSONValue {
    try await dispatcher.dispatch(name: "moot_file_fact", arguments: .object(arguments))
}

private func searchFacts(
    _ dispatcher: ToolDispatcher, _ arguments: [String: JSONValue]
) async throws -> JSONValue {
    try await dispatcher.dispatch(name: "moot_fact_search", arguments: .object(arguments))
}

private func factRows(_ result: JSONValue) -> [[String: JSONValue]] {
    result.objectValue?["structuredContent"]?.objectValue?["data"]?.objectValue?["facts"]?
        .arrayValue?.compactMap(\.objectValue) ?? []
}

/// Open a bare in-memory estate (no corpus, no vector store).
/// The dense lane is dark — this is the default mootx01 serve state when
/// no semantic wiring has been applied to the estate.
private func openBareEstate(identity: String = "aria-mcp-server")
    async throws -> (ToolDispatcher, GeniusLocusKit, EstateHandle)
{
    let kit = GeniusLocusKit()
    let owner = OwnerCredentials(ownerIdentifier: "fact-provenance-tests")
    let storage = InMemoryStorage(
        configuration: EstateConfiguration(estateID: UUID(), backend: .inMemory))
    _ = try await LocusKit.Estate.create(storage: storage, owner: owner)
    let handle = try await kit.open(storage: storage, owner: owner, identityKeyStore: InMemoryEstateIdentityKeyStore())
    // No corpus or vector store registered — dense lane dark.
    let dispatcher = ToolDispatcher(kit: kit, handle: handle, serverIdentity: identity)
    return (dispatcher, kit, handle)
}

// ---------------------------------------------------------------------------
// MARK: - Bug C: Host identity stamped on facts
// ---------------------------------------------------------------------------

@Suite("Bug C — Fact provenance identity injection", .serialized)
struct FactProvenanceIdentityTests {

    /// A dispatcher constructed with identity "mootx01" must stamp facts
    /// filed via moot_file_fact with addedBy="mootx01". With no explicit
    /// source_memory_id the fact is sourceless, so the source column is absent.
    ///
    /// Note: the S4 fact row (COMPOSER-02B §11.7) does not surface `addedBy`
    /// at the MCP layer — the field is stored in LocusKit's KGFact row and
    /// is verifiable at the storage layer, but it is not a rendered column.
    /// This test verifies: the fact is filed and retrievable, and the host
    /// identity does NOT appear as a source drawer ID in the row.
    @Test func factFiledWithMootx01IdentityGetsMootx01AddedBy() async throws {
        let (dispatcher, kit, handle) = try await openBareEstate(identity: "mootx01")
        defer { Task { try? await kit.close(handle) } }

        let fileResult = try await fileFact(dispatcher, [
            "subject": .string("Paris"),
            "predicate": .string("is_capital_of"),
            "object": .string("France"),
        ])
        let body = factText(of: fileResult)
        #expect(!body.isEmpty, "v2 file_fact must succeed; got: \(body)")

        // Retrieve the fact and verify it is in the S4 surface.
        let searchResult = try await searchFacts(dispatcher, ["query": .string("Paris")])
        let rows = factRows(searchResult)
        #expect(
            rows.contains { $0["subject"] == .string("Paris") },
            "fact filed via 'mootx01' identity must be retrievable; got: \(rows)"
        )
        // The host identity is never a source drawer ID; the sourceless fact
        // renders '-' in the source column. Verify identity contamination is absent.
        #expect(
            !rows.contains { $0["source_memory_id"] == .string("mootx01") },
            "host identity must not appear as a source memory id; got: \(rows)"
        )
    }

    /// A dispatcher constructed with identity "aria-mcp-server" must file facts
    /// that are retrievable via moot_fact_search. The S4 row (COMPOSER-02B §11.7)
    /// does not surface `addedBy` at the MCP layer (stored in LocusKit's KGFact
    /// row; not a rendered column). This test verifies the fact is filed and
    /// the identity does not contaminate the source column.
    @Test func factFiledWithAriaMcpIdentityGetsAriaMcpAddedBy() async throws {
        let (dispatcher, kit, handle) = try await openBareEstate(identity: "aria-mcp-server")
        defer { Task { try? await kit.close(handle) } }

        _ = try await fileFact(dispatcher, [
            "subject": .string("Berlin"),
            "predicate": .string("is_capital_of"),
            "object": .string("Germany"),
        ])

        let searchResult = try await searchFacts(dispatcher, ["query": .string("Berlin")])
        let rows = factRows(searchResult)
        #expect(
            rows.contains { $0["subject"] == .string("Berlin") },
            "fact filed via identity 'aria-mcp-server' must be retrievable; got: \(rows)"
        )
        // Host identity must not appear as a source drawer ID.
        #expect(
            !rows.contains { $0["source_memory_id"] == .string("aria-mcp-server") },
            "host identity must not appear as a source memory id; got: \(rows)"
        )
    }

    /// An explicit source_memory_id must name a drawer that exists in this estate.
    /// A fact inherits its source drawer's sensitivity, so an anchor that
    /// resolves to nothing is rejected rather than filed at the Normal
    /// default — filing it would disclose at a tier no drawer authorised.
    @Test func explicitSourceIdNamingNoDrawerFailsTheWrite() async throws {
        let (dispatcher, kit, handle) = try await openBareEstate(identity: "mootx01")
        defer { Task { try? await kit.close(handle) } }

        await #expect(throws: JSONRPCError.self) {
            _ = try await fileFact(dispatcher, [
                "subject": .string("Tokyo"),
                "predicate": .string("is_capital_of"),
                "object": .string("Japan"),
                "source_memory_id": .string("external-agent"),
            ])
        }

        // Nothing was filed, so the fact surface stays empty.
        let searchResult = try await searchFacts(dispatcher, ["query": .string("Tokyo")])
        let rows = factRows(searchResult)
        #expect(
            !rows.contains { $0["predicate"] == .string("is_capital_of") },
            "a rejected write must leave no fact behind; got: \(rows)"
        )
    }
}

// ---------------------------------------------------------------------------
// MARK: - Bug D: Dark-lane probe in moot_fact_search (COMPOSER-02B update)
// ---------------------------------------------------------------------------
// recall_provenance was removed from the payload per COMPOSER-02B (moved to
// log-side only). The dark-lane probe still runs when a query is supplied, but
// its result is not emitted into the text body. These tests verify the new
// contract: fact search completes without error on a dark estate, and no
// recall_provenance token appears in the payload in either the query or no-query
// path.

@Suite("Bug D — Fact search dark-lane probe (log-only)", .serialized)
struct FactSearchDarkLaneHintTests {

    /// When the dense lane is dark and a query is supplied, moot_fact_search
    /// must still return the matching facts successfully. The dark-lane probe
    /// runs internally but its output goes to the log, not the payload
    /// (recall_provenance removed from payload per COMPOSER-02B).
    @Test func factSearchSucceeds_whenQueryAndDenseLaneDark() async throws {
        let (dispatcher, kit, handle) = try await openBareEstate()
        defer { Task { try? await kit.close(handle) } }

        // File a fact so the estate is non-empty.
        _ = try await fileFact(dispatcher, [
            "subject": .string("Swift"),
            "predicate": .string("created_by"),
            "object": .string("Apple"),
        ])

        // Search with a query — dense lane is dark (no corpus), but the
        // probe is log-side only; the payload must still contain the fact.
        let result = try await searchFacts(dispatcher, ["query": .string("Swift")])
        let rows = factRows(result)
        #expect(
            rows.contains { $0["predicate"] == .string("created_by") },
            "moot_fact_search with a query must return matching facts; got: \(rows)"
        )
        // recall_provenance is log-side only — must NOT appear in the payload.
        let body = factText(of: result)
        #expect(
            !body.contains("recall_provenance:"),
            "recall_provenance must not appear in payload (log-side only); got: \(body)"
        )
    }

    /// When no query is supplied (returns all facts), no recall_provenance
    /// token is emitted — the dark-lane probe skips entirely.
    @Test func factSearchNoProvenanceHint_whenNoQuery() async throws {
        let (dispatcher, kit, handle) = try await openBareEstate()
        defer { Task { try? await kit.close(handle) } }

        _ = try await fileFact(dispatcher, [
            "subject": .string("Rust"),
            "predicate": .string("created_by"),
            "object": .string("Graydon Hoare"),
        ])

        // No query → list-all path → probe does not run → no recall_provenance.
        let result = try await searchFacts(dispatcher, [:])
        let body = factText(of: result)
        #expect(
            !body.contains("recall_provenance:"),
            "moot_fact_search without a query must NOT emit recall_provenance:; got: \(body)"
        )
    }
}

File: packages/kits/AriaMcpKit/rust/tests/fact_provenance_tests.rs
//! fact_provenance_tests.rs — Rust parallel tests for two fact-surface fixes.
//!
//! # Bug C — Host identity injected into EstateRegistry
//!
//! Every `moot_file_fact` call stamps `addedBy=<server_identity>` from the
//! registry field. `source=` carries the fact's local drawer anchor and is
//! empty when no `source_id` is supplied. The constant
//! `SERVER_ADDED_BY = "aria-mcp-server"` is gone; the correct identity is
//! injected at registry construction time and overrideable before the first
//! tool call. Tests verify:
//!
//! 1. A registry with `server_identity = "mootx01"` stamps `addedBy=mootx01`.
//! 2. A registry with the default identity stamps `addedBy=aria-mcp-server`.
//! 3. When the caller supplies an explicit `source_id`, that value wins over
//!    the registry identity.
//!
//! # Bug D — Dark-lane hint in moot_fact_search
//!
//! When the dense lane is dark (no corpus registered) and a query is supplied,
//! `moot_fact_search` appends a `recall_provenance:` line so AI callers can
//! distinguish "no lexical match" from "semantic search was not consulted".
//! Tests verify:
//!
//! 4. A bare estate (no corpus) + a query → `recall_provenance:` and
//!    `dense_lane:` tokens are present in the response.
//! 5. A bare estate (no corpus) with no query → `recall_provenance:` is
//!    absent (list-all path has no semantic query to misinterpret).
//!
//! Mirrors Swift `FactProvenanceTests.swift` in AriaMCPTests.

use std::collections::BTreeMap;

use aria_mcp::{
    dispatch::dispatch_tool,
    estate_registry::EstateRegistry,
    jsonrpc::JsonValue,
    surfaced_recall_ledger::SurfacedRecallLedger,
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

macro_rules! args {
    () => { BTreeMap::new() };
    ( $( $k:expr => $v:expr ),+ $(,)? ) => {{
        let mut m = BTreeMap::new();
        $( m.insert($k.to_string(), JsonValue::from(serde_json::json!($v))); )+
        m
    }};
}

fn content_text(result: &serde_json::Value) -> &str {
    result["content"][0]["text"].as_str().unwrap_or("")
}

fn is_success(result: &serde_json::Value) -> bool {
    result["isError"] == serde_json::json!(false)
}

// ---------------------------------------------------------------------------
// Bug C: Host identity stamped on facts
// ---------------------------------------------------------------------------

/// A registry with server_identity = "mootx01" must stamp facts with
/// addedBy=mootx01 when no explicit source_id is supplied.
///
/// Mirrors Swift test: factFiledWithMootx01IdentityGetsMootx01AddedBy.
#[test]
fn fact_filed_with_mootx01_identity_gets_mootx01_added_by() {
    // Bare estate — no corpus needed for provenance tests.
    let mut registry = EstateRegistry::new_inmemory_bare();
    registry.server_identity = "mootx01".to_owned();
    let ledger = SurfacedRecallLedger::new();

    let file_args = args![
        "subject" => "Paris",
        "predicate" => "is_capital_of",
        "object" => "France",
    ];
    let file_result = dispatch_tool("moot_file_fact", &file_args, &registry, &ledger)
        .expect("moot_file_fact must not error");
    assert!(
        is_success(&file_result),
        "moot_file_fact must succeed; got: {file_result:?}"
    );

    // Retrieve via moot_fact_search and inspect the addedBy= stamp.
    let search_args = args!["query" => "Paris"];
    let search_result = dispatch_tool("moot_fact_search", &search_args, &registry, &ledger)
        .expect("moot_fact_search must not error");
    let text = content_text(&search_result);
    assert!(
        text.contains("addedBy=mootx01"),
        "fact filed via identity 'mootx01' must carry addedBy=mootx01; got: {text}"
    );
    assert!(
        !text.contains("addedBy=aria-mcp-server"),
        "mootx01-hosted registry must NOT stamp 'aria-mcp-server'; got: {text}"
    );
}

/// A registry with the default server_identity ("aria-mcp-server") must stamp
/// facts with addedBy=aria-mcp-server.
///
/// Mirrors Swift test: factFiledWithAriaMcpIdentityGetsAriaMcpAddedBy.
#[test]
fn fact_filed_with_aria_mcp_identity_gets_aria_mcp_added_by() {
    // Set the host identity explicitly (mirrors Swift openBareEstate(identity:
    // "aria-mcp-server")). The bare registry defaults to "mootx01" (the product
    // identity), so this test — which verifies the aria-mcp-server identity path
    // — sets it rather than relying on a default.
    let mut registry = EstateRegistry::new_inmemory_bare();
    registry.server_identity = "aria-mcp-server".to_owned();
    let ledger = SurfacedRecallLedger::new();

    let file_args = args![
        "subject" => "Berlin",
        "predicate" => "is_capital_of",
        "object" => "Germany",
    ];
    dispatch_tool("moot_file_fact", &file_args, &registry, &ledger)
        .expect("moot_file_fact must not error");

    let search_args = args!["query" => "Berlin"];
    let search_result = dispatch_tool("moot_fact_search", &search_args, &registry, &ledger)
        .expect("moot_fact_search must not error");
    let text = content_text(&search_result);
    assert!(
        text.contains("addedBy=aria-mcp-server"),
        "fact filed via identity 'aria-mcp-server' must carry addedBy=aria-mcp-server; got: {text}"
    );
}

/// An explicit source_id must name a drawer that exists in this estate. A
/// fact inherits its source drawer's sensitivity, so an anchor that resolves
/// to nothing is rejected rather than filed at the Normal default — filing it
/// would disclose at a tier no drawer authorised.
///
/// Mirrors Swift test: explicitSourceIdNamingNoDrawerFailsTheWrite.
#[test]
fn explicit_source_id_naming_no_drawer_fails_the_write() {
    let mut registry = EstateRegistry::new_inmemory_bare();
    registry.server_identity = "mootx01".to_owned();
    let ledger = SurfacedRecallLedger::new();

    let file_args = args![
        "subject" => "Tokyo",
        "predicate" => "is_capital_of",
        "object" => "Japan",
        "source_id" => "external-agent",
    ];
    let file_result = dispatch_tool("moot_file_fact", &file_args, &registry, &ledger)
        .expect("moot_file_fact must return a tool result");
    assert!(
        !is_success(&file_result),
        "source_id naming no drawer must be rejected; got: {file_result:?}"
    );
    assert!(
        content_text(&file_result).contains("names no drawer"),
        "rejection must say why; got: {}",
        content_text(&file_result)
    );

    // Nothing was filed, so the fact surface stays empty.
    let search_args = args!["query" => "Tokyo"];
    let search_result = dispatch_tool("moot_fact_search", &search_args, &registry, &ledger)
        .expect("moot_fact_search must not error");
    let text = content_text(&search_result);
    assert!(
        !text.contains("Tokyo"),
        "a rejected write must leave no fact behind; got: {text}"
    );
}

// ---------------------------------------------------------------------------
// Bug D: Dark-lane hint in moot_fact_search
// ---------------------------------------------------------------------------

/// When the dense lane is dark (no corpus registered) and a query is supplied,
/// moot_fact_search must append a recall_provenance line so the caller knows
/// the match was lexical-only.
///
/// Mirrors Swift test: factSearchAppendsProvenance_whenQueryAndDenseLaneDark.
#[test]
fn fact_search_appends_provenance_when_query_and_dense_lane_dark() {
    // Bare estate — no corpus registered, dense lane is dark by design.
    let registry = EstateRegistry::new_inmemory_bare();
    let ledger = SurfacedRecallLedger::new();

    // File a fact so the estate is non-empty (proves the hint is about
    // lane state, not about the estate being empty).
    let file_args = args![
        "subject" => "Swift",
        "predicate" => "created_by",
        "object" => "Apple",
    ];
    dispatch_tool("moot_file_fact", &file_args, &registry, &ledger)
        .expect("moot_file_fact must not error");

    // Search with a query — dense lane is dark so a recall_provenance line
    // must appear.
    let search_args = args!["query" => "Swift"];
    let search_result = dispatch_tool("moot_fact_search", &search_args, &registry, &ledger)
        .expect("moot_fact_search must not error");
    let text = content_text(&search_result);
    assert!(
        text.contains("recall_provenance:"),
        "moot_fact_search with a query on a dark-dense-lane estate must emit recall_provenance:; got: {text}"
    );
    assert!(
        text.contains("dense_lane:"),
        "recall_provenance line must include a dense_lane: token; got: {text}"
    );
}

/// When no query is supplied (list-all path), no recall_provenance hint is
/// emitted — there is no semantic query to misinterpret.
///
/// Mirrors Swift test: factSearchNoProvenanceHint_whenNoQuery.
#[test]
fn fact_search_no_provenance_hint_when_no_query() {
    let registry = EstateRegistry::new_inmemory_bare();
    let ledger = SurfacedRecallLedger::new();

    let file_args = args![
        "subject" => "Rust",
        "predicate" => "created_by",
        "object" => "Graydon Hoare",
    ];
    dispatch_tool("moot_file_fact", &file_args, &registry, &ledger)
        .expect("moot_file_fact must not error");

    // No query → list-all path → no provenance hint.
    let search_args = args![];
    let search_result = dispatch_tool("moot_fact_search", &search_args, &registry, &ledger)
        .expect("moot_fact_search must not error");
    let text = content_text(&search_result);
    assert!(
        !text.contains("recall_provenance:"),
        "moot_fact_search without a query must NOT emit recall_provenance:; got: {text}"
    );
}
